### PR TITLE
[REF] website_slides: convert add course tag dialog to owl

### PR DESCRIPTION
**Pull request by lm4649**
As legacy dialog code will be removed soon, the following dialog shall be rewritten with OWL.

TagCourseDialog
=============

This dialog enables course managers to add or create course tag directly from the frontend.
You access it by clicking on 'Add Tag' on the top of the course page. It's a form with two dropdowns. The first one to select or create a tag. The second one is to assign the tag to a group and is displayed only when you want to create a new tag.
You can only add one tag at a time

Legacy  TagCourseDialog: see odoo/odoo#44365

task-3510874 / 3553728

---
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: addons/website_slides/static/src/js/public/components/course_tag_add_dialog/course_tag_add_dialog.js [hunk 1]
```diff
@@ -0,0 +1,233 @@
...
+        channelID: { type: Number, optional: true },
+        defaultTag: { type: String, optional: true },
+        tagIds: Array,
+        close: Function,
+    };
+    static template = "website_slides.CourseTagAddDialog";
+
+    setup() {
+        super.setup();
+        this.size = "md";
```
Comment (flch-odoo) [inline addons/website_slides/static/src/js/public/components/course_tag_add_dialog/course_tag_add_dialog.js:23]: If it is static, let's put it in the template directly.

File: addons/website_slides/static/src/js/public/components/course_tag_add_dialog/course_tag_add_dialog.js [hunk 2]
```diff
@@ -0,0 +1,233 @@
...
+            group_id: values.group_id,
+        });
+
+        if (data.error) {
+            this.state.alertDisplay = true;
+            this.state.alertMsg = data.error;
+        } else {
+            this.props.close();
+        }
+    }
```
Comment (flch-odoo) [inline addons/website_slides/static/src/js/public/components/course_tag_add_dialog/course_tag_add_dialog.js:140]: Can these be merged?

**Review by flch-odoo — COMMENTED** [review]
Validation comments here too :)

**Review by flch-odoo — COMMENTED** [review]
We're almost there here too :)

File: addons/website_slides/static/src/js/slides_course_tag_add.js [hunk 3]
```diff
@@ -343,8 +15,11 @@ publicWidget.registry.websiteSlidesTag = publicWidget.Widget.extend({
     //--------------------------------------------------------------------------
 
     _openDialog: function ($element) {
-        var data = $element.data();
-        return new TagCourseDialog(this, data).open();
+        const props = {
+            channelID: $element.data("channel-id"),
+            tagIds: $element.data("channel-tag-ids"),
+        };
+        this.call("dialog", "add", CourseTagAddDialog, props);
```
[the remark this answers is not in the document]
Comment (lm4649) [reply addons/website_slides/static/src/js/slides_course_tag_add.js]: My bad. Forgot to transfer _openDialog into _onAddTagClick.

**Review by lm4649 — COMMENTED** [review]
Changes done :+1:

File: addons/website_slides/static/src/js/slides_course_tag_add.js [hunk 4]
```diff
@@ -357,11 +19,14 @@ publicWidget.registry.websiteSlidesTag = publicWidget.Widget.extend({
      */
     _onAddTagClick: function (ev) {
         ev.preventDefault();
-        this._openDialog($(ev.currentTarget));
+        const props = {
```
Comment (flch-odoo) [inline addons/website_slides/static/src/js/slides_course_tag_add.js:22]: I'm still not convinced by this very short-lived variable. If you really like it though, we should at least give it a more explicit name like `addTagDialogProps` imho.

Comment (lm4649) [reply addons/website_slides/static/src/js/slides_course_tag_add.js:22]: ok, I replace props by an anoymous object.

File: addons/website_slides/static/src/js/public/components/course_tag_add_dialog/course_tag_add_dialog.js [hunk 5]
```diff
@@ -0,0 +1,232 @@
...
+            alertDisplay: false,
+            alertMsg: "",
+        });
+        this.validation = useState({
+            tagIsValid: false,
+            tagIsInvalid: false,
+            tagGroupIsValid: false,
+            tagGroupIsInvalid: false,
+        });
+        const tags = await this._fetch_choices("tag", [
```
Comment (std-odoo) [inline addons/website_slides/static/src/js/public/components/course_tag_add_dialog/course_tag_add_dialog.js:44]: the 2 `_fetch_choices` can be done in // with `await Promise.all`

File: addons/website_slides/static/src/js/public/components/course_tag_add_dialog/course_tag_add_dialog.js [hunk 6]
```diff
@@ -0,0 +1,232 @@
...
+    async _fetch_choices(type, domain = [], fields = ["name"]) {
+        const choices = [];
+        const { read_results, can_create } = await this.rpc(`/slides/channel/${type}/search_read`, {
+            fields,
+            domain,
+        });
+
+        read_results.forEach((choice) => {
+            choices.push({ value: choice.id, label: choice.name });
+        });
```
Comment (std-odoo) [inline addons/website_slides/static/src/js/public/components/course_tag_add_dialog/course_tag_add_dialog.js:194]: can we use map instead ?
```js
const choices = read_results.map((choice) => {
    return { value: choice.id, label: choice.name };
});
```

File: addons/website_slides/static/src/js/public/components/course_tag_add_dialog/course_tag_add_dialog.xml [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+            <a role="button" class="btn btn-primary me-1" t-on-click.prevent="onClickFormSubmit">
+                <span>Add</span>
+            </a>
+            <button type="button" class="btn btn-secondary" t-on-click="() => this.props.close()">
+                <span>Back</span>
+            </button>
+        </t>
+        <div class="mb-3">
+            <div t-if="state.alertDisplay" t-out="state.alertMsg" class="alert alert-warning"/>
+            <div t-att-class="{'mb-3': true, 'form-control is-valid': validation.tagIsValid, 'form-control is-invalid': validation.tagIsInvalid}">
```
Comment (std-odoo) [inline addons/website_slides/static/src/js/public/components/course_tag_add_dialog/course_tag_add_dialog.xml:16]: same below
```suggestion
            <div class="mb-3" t-att-class="{'form-control is-valid': validation.tagIsValid, 'form-control is-invalid': validation.tagIsInvalid}">

**Review by std-odoo — COMMENTED** [review]
@lm4649 Hi :) 

Good work overall, I have a few suggestions :) 

Cheers

**Review by lm4649 — COMMENTED** [review]
Thanks for the review and the improvements suggested :pray:

File: addons/website_slides/static/src/js/public/components/course_tag_add_dialog/course_tag_add_dialog.js [hunk 8]
```diff
@@ -0,0 +1,232 @@
...
+import { uniqueId } from "@web/core/utils/functions";
+import { useService } from "@web/core/utils/hooks";
+
+export class CourseTagAddDialog extends Component {
+    static components = { Dialog, DropdownItem, SelectMenu };
+    static props = {
+        channelID: { type: Number, optional: true },
+        defaultTag: { type: String, optional: true },
+        tagIds: Array,
+        close: Function,
```
[the remark this answers is not in the document]
Comment (std-odoo) [reply addons/website_slides/static/src/js/public/components/course_tag_add_dialog/course_tag_add_dialog.js:17]: I know, but will it be not flasy sometimes ? inwhic situation ? (can't find a place where we set that prop)

File: addons/website_slides/static/src/js/public/components/course_tag_add_dialog/course_tag_add_dialog.xml [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+                    choices="choices.tagGroupIds"
+                    onSelect.bind="onTagGroupSelect"
+                    required="true"
+                    togglerClass="'text-dark pe-1'"
+                    value="choices.tagGroupId"
+                >
+                    <t t-out="displayTagGroupValue"/>
+                    <t t-set-slot="bottomArea" t-slot-scope="select">
+                        <DropdownItem
+                                t-if="select.data.searchValue and state.canCreateTagGroup"
```
Comment (std-odoo) [inline addons/website_slides/static/src/js/public/components/course_tag_add_dialog/course_tag_add_dialog.xml:49]: still indentation issue here

File: addons/website_slides/static/src/js/public/components/course_tag_add_dialog/course_tag_add_dialog.xml [hunk 10]
```diff
@@ -0,0 +1,62 @@
...
+                            t-if="select.data.searchValue and state.canCreateTag"
+                            class="'btn text-primary'"
+                            onSelected="() => this.createChoice(select.data.searchValue)"
+                        >
+                            Create this tag "<i t-out="select.data.searchValue" />"
+                        </DropdownItem>
+                    </t>
+                </SelectMenu>
+            </div>
+            <div t-if="state.showTagGroup" t-att-class="{'mb-3': true, 'form-control is-valid': validation.tagGroupIsValid, 'form-control is-invalid': validation.tagGroupIsValid === false}">
```
Comment (std-odoo) [inline addons/website_slides/static/src/js/public/components/course_tag_add_dialog/course_tag_add_dialog.xml:37]: comment not solved here

image

**Review by std-odoo — COMMENTED** [review]
@lm4649 Hi :) 

I have a few more comments

Cheers

File: addons/website_slides/static/src/js/public/components/course_tag_add_dialog/course_tag_add_dialog.xml (hunk 10 above)
Comment (lm4649) [reply addons/website_slides/static/src/js/public/components/course_tag_add_dialog/course_tag_add_dialog.xml]: My bad :pray:

File: addons/website_slides/static/src/js/public/components/course_tag_add_dialog/course_tag_add_dialog.js (hunk 8 above)
[the remark this answers is not in the document]
Comment (lm4649) [reply addons/website_slides/static/src/js/public/components/course_tag_add_dialog/course_tag_add_dialog.js:17]: `close function` comes from the `DialogService` (dialog_service.js) used to add/remove this dialog. I have not encountered falsy cases so far and don't think one may get one.

close_props

**Review by lm4649 — COMMENTED** [review]
Thank for these additional comments and sorry for the omissions :pray:

**Comment by std-odoo** [comment]
@lm4649 Np :) About the close props, I just tested with the props removed and everything seems to work fine

File: addons/website_slides/static/src/js/public/components/course_tag_add_dialog/course_tag_add_dialog.js [hunk 11]
```diff
@@ -0,0 +1,229 @@
...
+            canCreateTagGroup: false,
+            canCreateTag: false,
+            alertMsg: "",
+        });
+        this.validation = useState({
+            tagIsValid: undefined,
+            tagGroupIsValid: undefined,
+        });
+        const [tags, groups] = await Promise.all([
+            this._fetch_choices("tag", [
```
Comment (std-odoo) [inline addons/website_slides/static/src/js/public/components/course_tag_add_dialog/course_tag_add_dialog.js:41]:
```suggestion
            this._fetchChoices("tag", [

Comment (lm4649) [reply addons/website_slides/static/src/js/public/components/course_tag_add_dialog/course_tag_add_dialog.js]: `About the close props, I just tested with the props removed and everything seems to work fine`:
Did you tried with the developer mode (with assets) mode on ?

**Review by lm4649 — COMMENTED** [review]
Hi, I camel cased the fetchChoices, but not sure we should remove the close props.

**Review by std-odoo — COMMENTED** [review]
@lm4649 Ah indeed, LGTM, thanks for your work :)

**Comment by std-odoo** [comment]
@robodoo r+

**Pull request closed without merging**